src/Features/Core/Portable/IntroduceVariable/AbstractIntroduceParameterService.cs

```diff
@@ -0,0 +1,862 @@
+        }
+
+        /// <summary>
+        /// Determines if the expression's grandparent is a variable declarator and if so,
+        /// returns the name
+        /// </summary>
+        protected static bool ShouldRemoveVariableDeclaratorContainingExpression(
+            Document document, TExpressionSyntax expression, [NotNullWhen(true)] out string? varDeclName, [NotNullWhen(true)] out SyntaxNode? localDeclaration)
+        {
+            var syntaxFacts = document.GetRequiredLanguageService<ISyntaxFactsService>();
+            localDeclaration = expression.FirstAncestorOrSelf<SyntaxNode>(node => syntaxFacts.IsLocalDeclarationStatement(node));
+            if (localDeclaration is null)
+            {
+                varDeclName = null;
+                return false;
+            }
+
+            if (syntaxFacts.GetVariablesOfLocalDeclarationStatement(localDeclaration).Count > 1)
+            {
+                varDeclName = null;
+                localDeclaration = null;
+                return false;
+            }
+
+            var expressionDecl = expression?.Parent?.Parent;
+            if (syntaxFacts.IsVariableDeclarator(expressionDecl))
+            {
+                varDeclName = syntaxFacts.GetIdentifierOfVariableDeclarator(expressionDecl).ValueText;
+                return true;
+            }
+
+            varDeclName = null;
+            localDeclaration = null;
+            return false;
+        }
+
+        /// <summary>
+        /// Locates all the call sites of the method that introduced the parameter
+        /// </summary>
+        protected static async Task<ImmutableDictionary<Document, List<TInvocationExpressionSyntax>>> FindCallSitesAsync(
+            Document document, IMethodSymbol methodSymbol, CancellationToken cancellationToken)
+        {
+            var methodCallSitesBuilder = ImmutableDictionary.CreateBuilder<Document, List<TInvocationExpressionSyntax>>();
+            var progress = new StreamingProgressCollector();
+            await SymbolFinder.FindReferencesAsync(
+                methodSymbol, document.Project.Solution, progress,
+                documents: null, FindReferencesSearchOptions.Default, cancellationToken).ConfigureAwait(false);
+            var referencedSymbols = progress.GetReferencedSymbols();
+
+            // Ordering by descending to sort invocations by starting span to account for nested invocations
+            var referencedLocations = referencedSymbols.SelectMany(referencedSymbol => referencedSymbol.Locations).Distinct()
+                .OrderByDescending(reference => reference.Location.SourceSpan.Start);
+
+            // Adding the original document to account for refactorings that do not have any invocations
+            methodCallSitesBuilder.Add(document, new List<TInvocationExpressionSyntax>());
+
+            foreach (var refLocation in referencedLocations)
+            {
+                // Does not support cross-language references currently
```

Comment:
do we have a test for this?

Reply:
No i need to find an example test that has multiple docs in different languages

Reply:
did you find this?  if not, i can show you examples.

Reply:
I did, added a test for it